File: typescript/shegerpay.ts
/**
 * ShegerPay TypeScript SDK
 * Official TypeScript/JavaScript SDK for ShegerPay Payment Verification Gateway
 * 
 * @example
 * import { ShegerPay } from '@shegerpay/sdk';
 * 
 * const client = new ShegerPay('sk_test_xxx');
 * const result = await client.verify({ transactionId: 'FT123456', amount: 100 });
 */

// ============================================
// Types & Interfaces
// ============================================

export interface ShegerPayConfig {
  apiKey: string;
  baseUrl?: string;
  timeout?: number;
}

export interface VerifyParams {
  transactionId: string;
  amount: number;
  provider?: 'cbe' | 'telebirr' | 'awash' | 'boa' | 'ebirr_kaafi' | 'ebirr_coop';
  merchantName?: string;
}

export interface VerificationResult {
  valid: boolean;
  status: 'verified' | 'failed' | 'pending' | 'error';
  provider?: string;
  transactionId?: string;
  amount?: number;
  payer?: string;
  receiver?: string;
  mode: 'test' | 'live';
  errorCode?: string;
  message?: string;
  suggestion?: string;
}

export interface PaymentLinkParams {
  title: string;
  amount: number;
  currency?: 'ETB' | 'USD';
  description?: string;
  enableCbe?: boolean;
  enableTelebirr?: boolean;
  enableCrypto?: boolean;
  expiresInHours?: number;
}

export interface PaymentLink {
  id: string;
  shortCode: string;
  url: string;
  qrCode: string;
  title: string;
  amount: number;
  currency: string;
  status: 'active' | 'inactive' | 'expired';
}

export interface WebhookParams {
  url: string;
  events: string[];
  secret?: string;
}

export interface Webhook {
  id: string;
  url: string;
  events: string[];
  secret: string;
  createdAt: string;
}

export interface CryptoPaymentParams {
  amountUsd: number;
  currency: 'USDT' | 'BTC' | 'ETH' | 'TRX';
  walletAddress: string;
  chain?: 'TRON' | 'ETH' | 'BSC' | 'BTC';
}

export interface CryptoPayment {
  referenceId: string;
  paymentAmount: string;
  currency: string;
  walletAddress: string;
  qrCode: string;
  expiresAt: string;
}

export interface TelegramConfig {
  botToken: string;
  chatId: string;
  notifyOnPayment?: boolean;
  notifyOnSecurity?: boolean;
}

export interface HealthStatus {
  status: 'healthy' | 'degraded' | 'unhealthy';
  version: string;
  timestamp: string;
  services: Record<string, string>;
  providers: Record<string, string>;
  encryption: {
    atRest: string;
    inTransit: string;
    apiKeys: string;
  };
}

export interface ProviderStatus {
  overallStatus: 'operational' | 'partial_outage' | 'major_outage';
  providersOnline: string;
  updatedAt: string;
  providers: Array<{
    id: string;
    name: string;
    status: 'online' | 'offline';
    avgResponseMs: number;
    successRate: number;
    lastSuccess: string;
  }>;
}

// ============================================
// Error Classes
// ============================================

export class ShegerPayError extends Error {
  statusCode: number;
  errorCode?: string;
  suggestion?: string;

  constructor(message: string, statusCode: number = 400, errorCode?: string, suggestion?: string) {
    super(message);
    this.name = 'ShegerPayError';
    this.statusCode = statusCode;
    this.errorCode = errorCode;
    this.suggestion = suggestion;
  }
}

export class AuthenticationError extends ShegerPayError {
  constructor(message: string = 'Invalid API key') {
    super(message, 401, 'AUTH_002');
    this.name = 'AuthenticationError';
  }
}

export class ValidationError extends ShegerPayError {
  constructor(message: string) {
    super(message, 400, 'VAL_001');
    this.name = 'ValidationError';
  }
}

// ============================================
// Main Client
// ============================================

export class ShegerPay {
  private apiKey: string;
  private baseUrl: string;
  private timeout: number;
  readonly mode: 'test' | 'live';

  constructor(apiKey: string, config?: Partial<ShegerPayConfig>) {
    if (!apiKey) {
      throw new AuthenticationError('API key is required');
    }
    
    if (!apiKey.startsWith('sk_test_') && !apiKey.startsWith('sk_live_')) {
      throw new AuthenticationError('Invalid API key format. Must start with sk_test_ or sk_live_');
    }

    this.apiKey = apiKey;
    this.baseUrl = (config?.baseUrl || 'https://api.shegerpay.com').replace(/\/$/, '');
    this.timeout = config?.timeout || 30000;
    this.mode = apiKey.startsWith('sk_test_') ? 'test' : 'live';
  }

  // ============================================
  // Ethiopian Payment Verification
  // ============================================

  /**
   * Verify a payment transaction
   */
  async verify(params: VerifyParams): Promise<VerificationResult> {
    const { transactionId, amount, provider, merchantName } = params;
    
    if (!transactionId) throw new ValidationError('transactionId is required');
    if (!amount) throw new ValidationError('amount is required');
    
    // Auto-detect provider
    const detectedProvider = provider || (transactionId.toUpperCase().startsWith('FT') ? 'cbe' : 'telebirr');
    
    return this.request<VerificationResult>('POST', '/api/v1/verify', {
      provider: detectedProvider,
      transaction_id: transactionId,
      amount,
      merchant_name: merchantName || 'ShegerPay Verification'
    });
  }

  /**
   * Quick verification with auto-detected provider
   */
  async quickVerify(transactionId: string, amount: number): Promise<VerificationResult> {
    return this.request<VerificationResult>('POST', '/api/v1/quick-verify', {
      transaction_id: transactionId,
      amount
    });
  }

  // ============================================
  // Payment Links
  // ============================================

  /**
   * Create a payment link
   */
  async createPaymentLink(params: PaymentLinkParams): Promise<PaymentLink> {
    const { title, amount, currency, description, enableCbe, enableTelebirr, enableCrypto, expiresInHours } = params;
    
    if (!title) throw new ValidationError('title is required');
    if (!amount) throw new ValidationError('amount is required');
    
    return this.request<PaymentLink>('POST', '/api/v1/payment-links/', {
      title,
      amount,
      currency: currency || 'ETB',
      description,
      enable_cbe: enableCbe !== false,
      enable_telebirr: enableTelebirr !== false,
      enable_crypto: enableCrypto || false,
      expires_in_hours: expiresInHours || 24
    });
  }

  /**
   * List all payment links
   */
  async listPaymentLinks(): Promise<PaymentLink[]> {
    const response = await this.request<{ links: PaymentLink[] }>('GET', '/api/v1/payment-links/');
    return response.links || [];
  }

  /**
   * Delete a payment link
   */
  async deletePaymentLink(linkId: string): Promise<{ success: boolean }> {
    return this.request('DELETE', `/api/v1/payment-links/${linkId}`);
  }

  // ============================================
  // Webhooks
  // ============================================

  /**
   * Create a webhook endpoint
   */
  async createWebhook(params: WebhookParams): Promise<Webhook> {
    const { url, events } = params;
    
    if (!url) throw new ValidationError('url is required');
    if (!events || events.length === 0) throw new ValidationError('events are required');
    
    return this.request<Webhook>('POST', '/api/v1/webhooks/', { url, events });
  }

  /**
   * List all webhooks
   */
  async listWebhooks(): Promise<Webhook[]> {
    return this.request<Webhook[]>('GET', '/api/v1/webhooks/');
  }

  /**
   * Delete a webhook
   */
  async deleteWebhook(webhookId: string): Promise<{ success: boolean }> {
    return this.request('DELETE', `/api/v1/webhooks/${webhookId}`);
  }

  /**
   * Test a webhook
   */
  async testWebhook(webhookId: string): Promise<{ success: boolean; statusCode: number }> {
    return this.request('POST', '/api/v1/webhooks/test', { webhook_id: webhookId });
  }

  // ============================================
  // Crypto Payments
  // ============================================

  /**
   * Get crypto prices
   */
  async getCryptoPrices(symbol?: string): Promise<Record<string, number>> {
    if (symbol) {
      return this.request('GET', `/api/v1/crypto/rate/${symbol.toUpperCase()}`);
    }
    return this.request('GET', '/api/v1/crypto/rates');
  }

  /**
   * Create a crypto payment intent
   */
  async createCryptoPayment(params: CryptoPaymentParams): Promise<CryptoPayment> {
    const { amountUsd, currency, walletAddress, chain } = params;
    
    if (!amountUsd) throw new ValidationError('amountUsd is required');
    if (!currency) throw new ValidationError('currency is required');
    if (!walletAddress) throw new ValidationError('walletAddress is required');
    
    return this.request<CryptoPayment>('POST', '/api/v1/crypto/generate-intent', {
      amount_usd: amountUsd,
      currency: currency.toUpperCase(),
      wallet_address: walletAddress,
      chain: chain || 'TRON'
    });
  }

  /**
   * Verify a crypto payment
   */
  async verifyCryptoPayment(referenceId: string, transactionHash?: string): Promise<VerificationResult> {
    if (!referenceId) throw new ValidationError('referenceId is required');
    
    return this.request<VerificationResult>('POST', '/api/v1/crypto/verify-reference', {
      reference_id: referenceId,
      transaction_hash: transactionHash
    });
  }

  // ============================================
  // Monitoring & Health
  // ============================================

  /**
   * Get detailed API health status
   */
  async getHealth(): Promise<HealthStatus> {
    return this.request<HealthStatus>('GET', '/api/v1/monitoring/health');
  }

  /**
   * Get provider status and uptime
   */
  async getProviderStatus(): Promise<ProviderStatus> {
    return this.request<ProviderStatus>('GET', '/api/v1/monitoring/providers');
  }

  /**
   * Get API usage metrics
   */
  async getMetrics(): Promise<Record<string, any>> {
    return this.request('GET', '/api/v1/monitoring/metrics');
  }

  /**
   * Get system uptime data
   */
  async getUptime(): Promise<Record<string, any>> {
    return this.request('GET', '/api/v1/monitoring/uptime');
  }

  // ============================================
  // Notifications
  // ============================================

  /**
   * Get notification settings
   */
  async getNotificationSettings(): Promise<Record<string, any>> {
    return this.request('GET', '/api/v1/notifications/settings');
  }

  /**
   * Configure Telegram notifications
   */
  async configureTelegram(config: TelegramConfig): Promise<{ success: boolean; testSent: boolean }> {
    const { botToken, chatId, notifyOnPayment, notifyOnSecurity } = config;
    
    if (!botToken) throw new ValidationError('botToken is required');
    if (!chatId) throw new ValidationError('chatId is required');
    
    return this.request('POST', '/api/v1/notifications/telegram/configure', {
      bot_token: botToken,
      chat_id: chatId,
      notify_on_payment: notifyOnPayment !== false,
      notify_on_security: notifyOnSecurity !== false
    });
  }

  /**
   * Send test Telegram notification
   */
  async testTelegram(): Promise<{ success: boolean; message: string }> {
    return this.request('POST', '/api/v1/notifications/telegram/test', {});
  }

  /**
   * Disable Telegram notifications
   */
  async disableTelegram(): Promise<{ success: boolean }> {
    return this.request('DELETE', '/api/v1/notifications/telegram');
  }

  // ============================================
  // Wallet & Transactions
  // ============================================

  /**
   * Get wallet balances
   */
  async getWalletBalance(): Promise<Record<string, number>> {
    return this.request('GET', '/api/v1/wallets/balances');
  }

  /**
   * Get transaction history
   */
  async listTransactions(filters?: { status?: string; provider?: string; limit?: number }): Promise<any[]> {
    const params = new URLSearchParams(filters as any).toString();
    return this.request('GET', `/api/v1/transactions/history${params ? '?' + params : ''}`);
  }

  /**
   * Get current subscription
   */
  async getSubscription(): Promise<Record<string, any>> {
    return this.request('GET', '/api/v1/subscriptions/current');
  }

  /**
   * Get API usage stats
   */
  async getUsage(): Promise<Record<string, any>> {
    return this.request('GET', '/api/v1/subscriptions/usage');
  }

  // ============================================
  // Internal Request Method
  // ============================================

  private async request<T>(method: string, path: string, body?: Record<string, any>): Promise<T> {
    const url = `${this.baseUrl}${path}`;
    
    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.apiKey}`,
      'Content-Type': 'application/json',
      'User-Agent': 'ShegerPay-TypeScript-SDK/1.0'
    };

    const options: RequestInit = {
      method,
      headers
    };

    if (body && method !== 'GET') {
      options.body = JSON.stringify(body);
    }

    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const response = await fetch(url, {
        ...options,
        signal: controller.signal
      });

      clearTimeout(timeoutId);

      if (response.status === 401) {
        throw new AuthenticationError('Invalid API key');
      }

      if (response.status === 400) {
        const error = await response.json();
        throw new ValidationError(error.detail || error.message || 'Validation error');
      }

      if (response.status >= 500) {
        throw new ShegerPayError('Server error', response.status, 'GEN_001');
      }

      return response.json() as Promise<T>;
    } catch (error: any) {
      if (error.name === 'AbortError') {
        throw new ShegerPayError('Request timed out', 408, 'GEN_002');
      }
      if (error instanceof ShegerPayError) {
        throw error;
      }
      throw new ShegerPayError(`Request failed: ${error.message}`, 500);
    }
  }
}

// ============================================
// Webhook Verification Helper
// ============================================

/**
 * Verify webhook signature
 */
export async function verifyWebhookSignature(
  payload: string,
  signature: string,
  secret: string
): Promise<boolean> {
  // Node.js environment
  if (typeof require !== 'undefined') {
    const crypto = require('crypto');
    const expected = crypto
      .createHmac('sha256', secret)
      .update(payload)
      .digest('hex');
    return `sha256=${expected}` === signature;
  }

  // Browser with Web Crypto API
  const encoder = new TextEncoder();
  const keyData = encoder.encode(secret);
  const data = encoder.encode(payload);

  const key = await crypto.subtle.importKey(
    'raw',
    keyData,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );

  const signatureBuffer = await crypto.subtle.sign('HMAC', key, data);
  const expected = Array.from(new Uint8Array(signatureBuffer))
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');

  return `sha256=${expected}` === signature;
}

// ============================================
// Exports
// ============================================

export default ShegerPay;

File: swift/ShegerPay.swift
/**
 * ShegerPay Swift SDK v2.2.0
 * Official Swift SDK for ShegerPay Payment Verification Gateway
 *
 * Usage:
 *   let client = ShegerPay(apiKey: "sk_test_xxx")
 *   let result = try await client.verify(transactionId: "FT123456", amount: 100, provider: "cbe")
 *   let imageResult = try await client.verifyImage("base64_or_url", provider: "cbe", amount: 100)
 */

import Foundation
import CommonCrypto

// MARK: - Errors

public enum ShegerPayError: Error, LocalizedError {
    case invalidApiKey
    case missingApiKey
    case authenticationFailed
    case validationError(String)
    case networkError(Error)
    case invalidResponse
    
    public var errorDescription: String? {
        switch self {
        case .invalidApiKey: return "Invalid API key format"
        case .missingApiKey: return "API key is required"
        case .authenticationFailed: return "Authentication failed"
        case .validationError(let msg): return msg
        case .networkError(let error): return error.localizedDescription
        case .invalidResponse: return "Invalid response from server"
        }
    }
}

// MARK: - Models

public struct VerificationResult: Codable {
    public let verified: Bool?
    public let valid: Bool
    public let status: String
    public let provider: String?
    public let transactionId: String?
    public let amount: Double?
    public let reason: String?
    public let mode: String?
    public let payer: String?
    
    enum CodingKeys: String, CodingKey {
        case verified, valid, status, provider, amount, reason, mode, payer
        case transactionId = "transaction_id"
    }
}

public struct PaymentLink: Codable {
    public let id: String
    public let shortCode: String
    public let paymentUrl: String
    public let qrCodeBase64: String
    public let status: String
    
    enum CodingKeys: String, CodingKey {
        case id, status
        case shortCode = "short_code"
        case paymentUrl = "payment_url"
        case qrCodeBase64 = "qr_code_base64"
    }
}

// MARK: - ShegerPay Client

public class ShegerPay {
    private let apiKey: String
    private let baseURL: String
    private let mode: String
    private let session: URLSession
    
    public init(apiKey: String, baseURL: String = "https://api.shegerpay.com") throws {
        guard !apiKey.isEmpty else {
            throw ShegerPayError.missingApiKey
        }
        
        guard apiKey.hasPrefix("sk_test_") || apiKey.hasPrefix("sk_live_") else {
            throw ShegerPayError.invalidApiKey
        }
        
        self.apiKey = apiKey
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.mode = apiKey.hasPrefix("sk_test_") ? "test" : "live"
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }
    
    // MARK: - Verification
    
    /// Verify a payment transaction
    public func verify(
        transactionId: String,
        amount: Double,
        provider: String? = nil,
        merchantName: String? = nil,
        senderAccount: String? = nil
    ) async throws -> VerificationResult {
        let detectedProvider = provider ?? (transactionId.lowercased().contains("cs.bankofabyssinia.com/slip/?trx=") ? "boa" : nil)
        guard let detectedProvider else {
            throw ShegerPayError.validationError("provider is required for ambiguous transaction references. Pass provider explicitly or use quickVerify().")
        }
        
        var params: [String: String] = [
            "provider": detectedProvider,
            "transaction_id": transactionId,
            "amount": String(amount),
            "merchant_name": merchantName ?? "ShegerPay Verification"
        ]
        if let senderAccount, !senderAccount.isEmpty {
            params["sender_account"] = senderAccount
        }
        
        return try await request(method: "POST", path: "/api/v1/verify", params: params)
    }
    
    /// Quick verification with auto-detected provider
    public func quickVerify(
        transactionId: String,
        amount: Double,
        expectedProvider: String? = nil,
        senderAccount: String? = nil
    ) async throws -> VerificationResult {
        var params: [String: String] = [
            "transaction_id": transactionId,
            "amount": String(amount)
        ]
        if let expectedProvider, !expectedProvider.isEmpty {
            params["expected_provider"] = expectedProvider
        }
        if let senderAccount, !senderAccount.isEmpty {
            params["sender_account"] = senderAccount
        }
        return try await request(method: "POST", path: "/api/v1/quick-verify", params: params)
    }
    
    // MARK: - Image Verification

    /// Verify payment from a receipt screenshot (base64 encoded string or public URL)
    public func verifyImage(
        _ image: String,
        provider: String? = nil,
        amount: Double? = nil,
        merchantName: String = "ShegerPay Verification"
    ) async throws -> VerificationResult {
        var params: [String: Any] = ["image": image, "merchant_name": merchantName]
        if let provider, !provider.isEmpty { params["provider"] = provider }
        if let amount { params["amount"] = amount }
        return try await requestJSON(method: "POST", path: "/api/v1/verify/image", json: params)
    }

    // MARK: - Payment Links
    
    /// Create a payment link
    public func createPaymentLink(
        title: String,
        amount: Double,
        currency: String = "ETB",
        description: String? = nil
    ) async throws -> PaymentLink {
        var params: [String: Any] = [
            "title": title,
            "amount": amount,
            "currency": currency,
            "enable_cbe": true,
            "enable_telebirr": true
        ]
        
        if let desc = description {
            params["description"] = desc
        }
        
        return try await requestJSON(method: "POST", path: "/api/v1/payment-links/", json: params)
    }

    // MARK: - Promo Codes

    public func createPromoCode(_ params: [String: Any]) async throws -> [String: Any] {
        try await requestDictionary(method: "POST", path: "/api/v1/promo-codes/", json: promoPayload(params))
    }

    public func listPromoCodes() async throws -> [[String: Any]] {
        try await requestArray(method: "GET", path: "/api/v1/promo-codes/", json: nil)
    }

    public func updatePromoCode(_ codeId: String, params: [String: Any]) async throws -> [String: Any] {
        try await requestDictionary(method: "PATCH", path: "/api/v1/promo-codes/\(codeId)", json: promoPayload(params))
    }

    public func deletePromoCode(_ codeId: String) async throws -> [String: Any] {
        try await requestDictionary(method: "DELETE", path: "/api/v1/promo-codes/\(codeId)", json: nil)
    }

    public func validatePromoCode(code: String, amount: Double, options: [String: Any] = [:]) async throws -> [String: Any] {
        var body = options
        body["code"] = code
        body["amount"] = amount
        return try await requestDictionary(method: "POST", path: "/api/v1/promo-codes/validate", json: body)
    }

    public func redeemPromoCode(code: String, amount: Double, transactionId: String, options: [String: Any] = [:]) async throws -> [String: Any] {
        var body = options
        body["code"] = code
        body["amount"] = amount
        body["transaction_id"] = transactionId
        return try await requestDictionary(method: "POST", path: "/api/v1/promo-codes/redeem", json: body)
    }

    public func applyPaymentLinkCoupon(shortCode: String, code: String, amount: Double? = nil, quantity: Int = 1, provider: String? = nil, customerIdentifier: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["code": code, "quantity": quantity]
        if let amount { body["amount"] = amount }
        if let provider { body["provider"] = provider }
        if let customerIdentifier { body["customer_identifier"] = customerIdentifier }
        return try await requestDictionary(method: "POST", path: "/api/v1/payment-links/\(shortCode)/apply-coupon", json: body)
    }

    public func getPaymentLinkOrderStatus(shortCode: String, orderId: String) async throws -> [String: Any] {
        return try await requestDictionary(
            method: "GET",
            path: "/api/v1/payment-links/\(shortCode)/orders/\(orderId)/status",
            json: nil
        )
    }
    
    // MARK: - Private Methods
    
    private func request<T: Decodable>(
        method: String,
        path: String,
        params: [String: String]
    ) async throws -> T {
        var urlComponents = URLComponents(string: baseURL + path)!
        
        var request = URLRequest(url: urlComponents.url!)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("ShegerPay-Swift-SDK/2.2.0", forHTTPHeaderField: "User-Agent")
        
        if method == "POST" {
            let body = params.map { "\($0.key)=\($0.value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? $0.value)" }
                .joined(separator: "&")
            request.httpBody = body.data(using: .utf8)
        }
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ShegerPayError.invalidResponse
        }
        
        if httpResponse.statusCode == 401 {
            throw ShegerPayError.authenticationFailed
        }
        
        return try JSONDecoder().decode(T.self, from: data)
    }
    
    private func requestJSON<T: Decodable>(
        method: String,
        path: String,
        json: [String: Any]
    ) async throws -> T {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: json)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ShegerPayError.invalidResponse
        }
        
        if httpResponse.statusCode == 401 {
            throw ShegerPayError.authenticationFailed
        }
        
        return try JSONDecoder().decode(T.self, from: data)
    }

    private func requestDictionary(
        method: String,
        path: String,
        json: [String: Any]?
    ) async throws -> [String: Any] {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let json, method != "GET", method != "DELETE" {
            request.httpBody = try JSONSerialization.data(withJSONObject: json)
        }

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ShegerPayError.invalidResponse
        }
        if httpResponse.statusCode == 401 {
            throw ShegerPayError.authenticationFailed
        }
        if httpResponse.statusCode == 204 {
            return [:]
        }
        if !(200...299).contains(httpResponse.statusCode) {
            throw ShegerPayError.validationError(String(data: data, encoding: .utf8) ?? "ShegerPay request failed")
        }
        return (try JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
    }

    private func requestArray(
        method: String,
        path: String,
        json: [String: Any]?
    ) async throws -> [[String: Any]] {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let json, method != "GET", method != "DELETE" {
            request.httpBody = try JSONSerialization.data(withJSONObject: json)
        }

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ShegerPayError.invalidResponse
        }
        if httpResponse.statusCode == 401 {
            throw ShegerPayError.authenticationFailed
        }
        if !(200...299).contains(httpResponse.statusCode) {
            throw ShegerPayError.validationError(String(data: data, encoding: .utf8) ?? "ShegerPay request failed")
        }
        return (try JSONSerialization.jsonObject(with: data)) as? [[String: Any]] ?? []
    }

    private func promoPayload(_ params: [String: Any]) -> [String: Any] {
        Dictionary(uniqueKeysWithValues: params.map { (snakeCase($0.key), $0.value) })
    }

    private func snakeCase(_ value: String) -> String {
        value.reduce(into: "") { output, character in
            if character.isUppercase {
                output.append("_")
                output.append(character.lowercased())
            } else {
                output.append(character)
            }
        }
    }
    
    // MARK: - Webhook Verification
    
    /// Verify webhook signature
    public static func verifyWebhookSignature(payload: String, signature: String, secret: String) -> Bool {
        guard let keyData = secret.data(using: .utf8),
              let payloadData = payload.data(using: .utf8) else {
            return false
        }
        
        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        
        keyData.withUnsafeBytes { keyBytes in
            payloadData.withUnsafeBytes { dataBytes in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256),
                       keyBytes.baseAddress, keyData.count,
                       dataBytes.baseAddress, payloadData.count,
                       &digest)
            }
        }
        
        let expected = "sha256=" + digest.map { String(format: "%02x", $0) }.joined()
        return expected == signature
    }

    public static func verifyRedirectSignature(params: [String: Any], signature: String, secret: String) -> Bool {
        let amountValue = Double("\(params["amount"] ?? "0")") ?? 0
        let amount = String(format: "%.2f", amountValue)
        let payload = [
            "\(params["checkout_session_id"] ?? params["checkoutSessionId"] ?? "")",
            "\(params["order_id"] ?? params["orderId"] ?? "")",
            "\(params["short_code"] ?? params["shortCode"] ?? "")",
            amount,
            "\(params["currency"] ?? "ETB")",
            "\(params["status"] ?? "paid")"
        ].joined(separator: "|")
        guard let keyData = secret.data(using: .utf8),
              let payloadData = payload.data(using: .utf8) else {
            return false
        }

        var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        keyData.withUnsafeBytes { keyBytes in
            payloadData.withUnsafeBytes { dataBytes in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256),
                       keyBytes.baseAddress, keyData.count,
                       dataBytes.baseAddress, payloadData.count,
                       &digest)
            }
        }
        let expected = digest.map { String(format: "%02x", $0) }.joined()
        return expected == signature.replacingOccurrences(of: "sha256=", with: "")
    }
}
